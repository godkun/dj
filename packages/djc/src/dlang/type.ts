import { IType } from '../types'
import { Entity } from './entity'

// ~~~~~~~~~ Boolean && boolean ~~~~~~~~~~~~~~~
export function Boolean() {
  return {
    tsType: 'boolean',
    javaType: 'java.Boolean',
  }
}

export function boolean() {
  return {
    tsType: 'boolean',
    javaType: 'java.boolean',
  }
}

// ~~~~~~~~~ each number type ~~~~~~~~~~~~~~
export function Integer() {
  return {
    tsType: 'number',
    javaType: 'java.Integer',
  }
}

export function int() {
  return {
    tsType: 'number',
    javaType: 'java.int',
  }
}

export function short() {
  return {
    tsType: 'number',
    javaType: 'java.short',
  }
}

export function Short(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Short',
  }
}

export function byte(): IType {
  return {
    tsType: 'number',
    javaType: 'java.byte',
  }
}

export function Byte(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Byte',
  }
}

export function long(): IType {
  return {
    tsType: 'number',
    javaType: 'java.long',
  }
}

export function Long(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Long',
  }
}

export function double(): IType {
  return {
    tsType: 'number',
    javaType: 'java.double',
  }
}

export function Double(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Double',
  }
}

export function float(): IType {
  return {
    tsType: 'number',
    javaType: 'java.float',
  }
}

export function Float(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Float',
  }
}

// ~~~~~~~~~~~~~~ String ~~~~~~~~~~~~~~~~
export function String(): IType {
  return {
    tsType: 'string',
    javaType: 'java.String',
  }
}

export function char(): IType {
  return {
    tsType: 'string',
    javaType: 'java.char',
  }
}

export function Char(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Char',
  }
}

export function chars(): IType {
  return {
    tsType: 'string',
    javaType: 'java.chars',
  }
}

export function Character(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Character',
  }
}

// ~~~~~~~~~~~~~~~~~~~~~ container ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export function List(type: Entity | (() => IType)): () => IType {
  const generic = type instanceof Entity ? type : type()
  return () => ({
    tsType: 'Array',
    javaType: 'java.List',
    generic: [generic],
  })
}

export function Set(type: Entity | (() => IType)): () => IType {
  const generic = type instanceof Entity ? type : type()
  return () => ({
    tsType: 'Array',
    javaType: 'java.Set',
    generic: [generic],
  })
}

export function Collection(type: Entity | (() => IType)): () => IType {
  const generic = type instanceof Entity ? type : type()
  return () => ({
    tsType: 'Array',
    javaType: 'java.Collection',
    generic: [generic],
  })
}

export function Iterator(type: Entity | (() => IType)): () => IType {
  const generic = type instanceof Entity ? type : type()
  return () => ({
    tsType: 'Array',
    javaType: 'java.Iterator',
    generic: [generic],
  })
}

export function Enumeration(type: Entity | (() => IType)): () => IType {
  const generic = type instanceof Entity ? type : type()
  return () => ({
    tsType: 'Array',
    javaType: 'java.Iterator',
    generic: [generic],
  })
}

export function HashMap(
  leftType: Entity | (() => IType),
  rightType: Entity | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.HashMap',
    generic: [
      leftType instanceof Entity ? leftType : leftType(),
      rightType instanceof Entity ? rightType : rightType(),
    ],
  })
}

export function Map(
  leftType: Entity | (() => IType),
  rightType: Entity | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.Map',
    generic: [
      leftType instanceof Entity ? leftType : leftType(),
      rightType instanceof Entity ? rightType : rightType(),
    ],
  })
}

export function Dictionary(
  leftType: Entity | (() => IType),
  rightType: Entity | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.Dictionary',
    generic: [
      leftType instanceof Entity ? leftType : leftType(),
      rightType instanceof Entity ? rightType : rightType(),
    ],
  })
}

export function Currency(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Currency',
  }
}
