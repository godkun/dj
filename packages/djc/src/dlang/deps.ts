type fullClassName = string

export interface ICount {
  offset: number
  map: {
    [k in fullClassName]: {
      clsName: string
      isDefault: boolean
    }
  }
}

/**
 * collect deps
 */
export default class Deps {
  private map: Map<string, ICount> = new Map()

  add(fullClsName: string, clsName: string, isDefault: boolean = true) {
    let val = this.map.get(clsName)

    if (val) {
      // 如果已经存在
      const existClsName = val.map[fullClsName]
      if (existClsName) {
        return existClsName.clsName
      }

      // 不存在
      const renameClsName = `${clsName}${val.offset}`
      val.map[fullClsName] = {
        clsName: renameClsName,
        isDefault,
      }
      val.offset++
      return renameClsName
    } else {
      this.map.set(clsName, {
        offset: 1,
        map: {
          [fullClsName]: {
            clsName,
            isDefault,
          },
        },
      })
    }

    return clsName
  }

  get imports() {
    const imports = []
    for (let { map } of this.map.values()) {
      for (let [fullClassName, { clsName, isDefault }] of Object.entries(map)) {
        imports.push({
          fullClassName,
          importName: clsName,
          isDefault,
        })
      }
    }
    return imports
  }
}
