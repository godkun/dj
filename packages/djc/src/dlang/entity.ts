import { IEntityField, IType } from '../types'
import Lang from './lang'

export class Entity extends Lang {
  fields: Array<IEntityField> = []
  constructor(fullClsName: string, comment?: string) {
    super(fullClsName, comment)
    this.deps.add('js-to-java', 'java')
  }

  field(name: string, type: Entity | (() => IType), comment?: string) {
    if (type instanceof Entity) {
      this.javaClazzFiled(type, name, comment)
      return this
    }

    // 获取当前类型信息
    const { tsType, javaType, generic } = type()

    // 当前field为基础类型，不带泛型的
    if (!generic || generic.length === 0) {
      this.javaBasicTypeField(name, tsType, javaType, comment)
      return this
    }

    // 容器类型 带泛型
    // 如List,Set带一种类型
    // Map, HashMap等两种类型

    // add suger transform
    this.deps.add('@dubbo/dj-suger', 's')

    // 针对泛型的处理 - 一元泛型- 针对list, Set, Collection, Iterator
    if (generic.length === 1) {
      // 获取当前的泛型的类型
      this.oneGenericField(generic, name, tsType, javaType, comment)
      return this
    }

    // 针对二元泛型的处理， Map, HashMap, Dictionary
    else if (generic.length === 2) {
      this.twoGenericsField(generic, name)
    }

    return this
  }

  private twoGenericsField(generic: (Entity | IType)[], name: string) {
    const [lg, rg] = generic

    // 二元的左子树泛型 只支持java.String
    if (lg instanceof Entity || lg.javaType !== 'java.String') {
      throw new Error(
        'Map/HashMap/Dictionary left generic type only support java.String'
      )
    }

    // 右侧泛型是 JavaClazz类型
    if (rg instanceof Entity) {
      const renamed = this.deps.add(rg.fullClsName, rg.clsName)
      this.fields.push({
        name,
        type: {
          tsType: `Map<string, ${renamed}>`,
          javaType: `s.$mhs(this.${name})`,
        },
        comment: '',
      })
    }

    // 右侧类型为基本数据类型
    else {
      this.fields.push({
        name,
        type: {
          tsType: `Map<string, ${rg.tsType}>`,
          javaType: `s.$mhs(this.${name}, ${rg.javaType})`,
        },
        comment: '',
      })
    }
  }

  private oneGenericField(
    generic: (Entity | IType)[],
    name: string,
    tsType: string,
    javaType: string,
    comment: string | undefined
  ) {
    const g = generic[0]
    // 如果当前的泛型类型是 JavaClass
    if (g instanceof Entity) {
      const renamed = this.deps.add(g.fullClsName, g.clsName)
      this.fields.push({
        name,
        type: {
          tsType: `${tsType}<${renamed}>`,
          javaType: `${javaType}(s.$lhs(this.${name}))`,
        },
        comment: this.fieldComment(comment),
      })
    } else {
      // 泛型的类型为基本类型
      this.fields.push({
        name,
        type: {
          tsType: `${tsType}<${g.tsType}>`,
          javaType: `${javaType}(s.$lhs(this.${name}, ${g.javaType}))`,
        },
        comment: this.fieldComment(comment),
      })
    }
  }

  private javaBasicTypeField(
    name: string,
    tsType: string,
    javaType: string,
    comment: string | undefined
  ) {
    this.fields.push({
      name,
      type: {
        tsType: tsType,
        javaType: `${javaType}(this.${name})`,
      },
      comment: this.fieldComment(comment),
    })
  }

  private javaClazzFiled(
    type: Entity,
    name: string,
    comment: string | undefined
  ) {
    const renamed = this.deps.add(type.fullClsName, type.clsName)
    this.fields.push({
      name,
      type: {
        tsType: renamed,
        javaType: `this.${name}.__fields2java()`,
      },
      comment: this.fieldComment(comment),
    })
  }
}

export default function entity(fullClsName: string, comment?: string) {
  const e = new Entity(fullClsName, comment)
  return {
    field(name: string, type: Entity | (() => IType), comment?: string) {
      e.field(name, type, comment)
      return this
    },
    ok() {
      return e
    },
  }
}
