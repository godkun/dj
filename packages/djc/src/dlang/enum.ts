import Lang from './lang'

export class Enum extends Lang {
  fields: Array<{ name: string; val: number | string; comment: string }> = []

  constructor(fullClsName: string, comment?: string) {
    super(fullClsName, comment)
  }

  field(name: string, val: number | string, comment?: string) {
    this.fields.push({
      name,
      val,
      comment: this.fieldComment(comment),
    })
    return this
  }
}

class EnumBuilder {
  constructor(private e: Enum) {}

  field(name: string, val: number | string, comment?: string) {
    this.e.field(name, val, comment)
    return this
  }

  ok() {
    return this.e
  }
}

export default function enumer(cls: string, comment?: string) {
  const e = new Enum(cls, comment)
  return new EnumBuilder(e)
}
