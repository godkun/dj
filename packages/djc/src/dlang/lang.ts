import Deps from './deps'

export default class Lang {
  clsName: string
  infName: string
  deps: Deps

  constructor(public fullClsName: string) {
    this.clsName = fullClsName.split('.').pop() || ''
    this.infName = `I${this.clsName}`
    this.deps = new Deps()
  }

  getComment(comment?: string) {
    return comment
      ? `
    /**
    * ${comment}
    */
   `
      : ''
  }
}
