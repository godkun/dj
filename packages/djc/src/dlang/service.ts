import { IType, TServiceMethod } from '../types'
import Lang from './lang'
import { Entity } from './entity'
import { Enum } from './enum'

export class Service extends Lang {
  _group: string = ''
  _version: string = '1.0.0'
  _methods: TServiceMethod = {}
  private _curMethodName: string = ''

  constructor(fullClsName: string, comment?: string) {
    super(fullClsName, comment)
  }

  group(group: string) {
    this._group = group
    return this
  }

  version(version: string) {
    this._version = version
    return this
  }

  method(name: string) {
    // set current method
    this._curMethodName = name

    // init method
    this._methods[name] = {
      args: [],
      ret: null,
    } as any

    return this
  }

  arg(name: string, type: Entity | Enum | (() => IType)) {
    const method = this._methods[this._curMethodName]

    if (type instanceof Enum) {
      const renamed = this.deps.add(type.fullClsName, type.clsName)
      method.args.push({ name, type: renamed })
      return this
    }

    if (type instanceof Entity) {
      const renamed = this.deps.add(type.fullClsName, type.infName, false)
      method.args.push({
        name,
        type: renamed,
      })
      return this
    }

    // 获取当前的类型信息
    const { tsType, generic } = type()

    // 如果当前类型没有泛型
    if (!generic || generic.length === 0) {
      method.args.push({
        name,
        type: tsType,
      })

      return this
    }

    // 如果当前有泛型，泛型参数有一个

    if (generic?.length === 1) {
      const g = generic[0]
      if (g instanceof Entity) {
        const renamed = this.deps.add(g.fullClsName, g.infName)
        method.args.push({
          name,
          type: `${tsType}<${renamed}>`,
        })
      } else if (g instanceof Enum) {
        const rename = this.deps.add(g.fullClsName, g.clsName)
        method.args.push({ name, type: `${tsType}<${rename}>` })
      } else {
        method.args.push({
          name,
          type: `${tsType}<${g.tsType}>`,
        })
      }
      return this
    }

    if (generic.length === 2) {
      const [lg, rg] = generic
      if (lg instanceof Entity || lg.javaType !== 'java.String') {
        throw new Error(
          'Map/HashMap/Dictionary left generic type only support java.String'
        )
      }
      if (rg instanceof Entity) {
        const renamed = this.deps.add(rg.fullClsName, rg.infName)
        method.args.push({
          name,
          type: `${tsType}<string, ${renamed}`,
        })
      } else if (rg instanceof Enum) {
        const renamed = this.deps.add(rg.fullClsName, rg.clsName)
        method.args.push({
          name,
          type: `${tsType}<string, ${renamed}`,
        })
      } else {
        method.args.push({
          name,
          type: `${tsType}<string, ${rg.tsType}>`,
        })
      }
    }

    return this
  }

  ret(type: Entity | Enum | (() => IType)) {
    // get current method meta data
    const method = this._methods[this._curMethodName]

    if (type instanceof Enum) {
      const renamed = this.deps.add(type.fullClsName, type.clsName)
      method.ret = renamed
      return this
    }

    if (type instanceof Entity) {
      const renameClsName = this.deps.add(type.fullClsName, type.clsName)
      method.ret = renameClsName
      return this
    }

    const { tsType, generic } = type()

    // basic type, not generic type
    if (!generic || generic.length === 0) {
      method.ret = tsType
      return this
    }

    // 如果只有一个泛型
    if (generic.length === 1) {
      const g = generic[0]
      if (g instanceof Entity) {
        const rename = this.deps.add(g.fullClsName, g.infName)
        method.ret = `${tsType}<${rename}>`
      } else if (g instanceof Enum) {
        const rename = this.deps.add(g.fullClsName, g.clsName)
        method.ret = `${tsType}<${rename}>`
      } else {
        method.ret = `${tsType}<${g.tsType}>`
      }
      return this
    }

    // 如果有两个泛型
    if (generic.length === 2) {
      const [lg, rg] = generic
      if (lg instanceof Entity || lg.javaType !== 'java.String') {
        throw new Error(
          'Map/HashMap/Dictionary left generic type only support java.String'
        )
      }

      if (rg instanceof Entity) {
        const rename = this.deps.add(rg.fullClsName, rg.infName)
        method.ret = `${tsType}<string, ${rename}>`
      } else if (rg instanceof Enum) {
        const rename = this.deps.add(rg.fullClsName, rg.clsName)
        method.ret = `${tsType}<string, ${rename}>`
      } else {
        method.ret = `${tsType}<string, ${rg.tsType}>`
      }
    }

    return this
  }
}

export default function service(fullClsName: string, comment?: string) {
  const s = new Service(fullClsName, comment)
  return {
    group(group: string) {
      s.group(group)
      return this
    },

    version(version: string) {
      s.version(version)
      return this
    },

    method(name: string) {
      s.method(name)
      return this
    },

    arg(name: string, type: Entity | Enum | (() => IType)) {
      s.arg(name, type)
      return this
    },

    ret(type: Entity | Enum | (() => IType)) {
      s.ret(type)
      return this
    },

    ok() {
      return s
    },
  }
}
