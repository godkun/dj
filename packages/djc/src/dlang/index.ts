/**
 * dl: dubbo lang
 *
 * 通过设计fluent javascript api来构建一个DSL style的代码描述language
 *
 * entity => object entity => java/js [class]
 *                         => go struct
 *
 * service => dubbo service
 *
 * enumer => js/java/go => enumeration
 *
 */

import entity from './entity'
import enumer from './enum'
import service from './service'
import {
  Integer,
  String,
  Boolean,
  List,
  Set,
  Collection,
  Map,
  boolean,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  char,
  Char,
  chars,
  Character,
  Enumeration,
  Iterator,
  Currency,
  Dictionary,
  HashMap,
} from './type'

export {
  // model entity
  entity,
  service,
  enumer,
  // data type
  Boolean,
  boolean,
  Integer,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  String,
  char,
  Char,
  chars,
  Character,
  // container data type
  List,
  Set,
  Collection,
  Iterator,
  Enumeration,
  Map,
  HashMap,
  Dictionary,
  Currency,
}
