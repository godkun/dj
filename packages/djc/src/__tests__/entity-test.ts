import * as dl from '../dlang'
import {
  AbstractServiceEmitter,
  ConsumerEmitter,
  EntityEmitter,
  fmt,
  ServiceEmitter,
} from '../emitter'
import EnumEmitter from '../emitter/enum'
import AbstractService from '../emitter/service-abstract'

const color = dl
  .enumer('org.apache.dubbo.enum.Color', '颜色类型的枚举')
  .field('RED', 0)
  .field('GREEN', 1)
  .field('BLUE', 2)
  .ok()

const user = dl
  .entity('org.apache.dubbo.domain.User', '用户模块')
  .field('id', dl.Integer, '用户id')
  .field('name', dl.String, '用户名')
  .field('color', color)
  .ok()

const userService = dl
  .service('org.apache.dubbo.service.UserService', '用户服务')
  .group('dubbo')
  .version('1.0.0')
  .method('sayHello')
  .arg('name', dl.String)
  .ret(dl.String)
  .ok()

const enumService = dl
  .service('org.apache.dubbo.service.EnumService')
  .group('dubbo')
  .version('1.0.0')
  .method('sayHello')
  .arg('color', color)
  .ret(color)
  .ok()

const gService = dl
  .service('org.apache.dubbo.service.GenService')
  .group('dubbo')
  .version('1.0.0')
  .method('sayHello')
  .arg('name', dl.List(dl.String))
  .method('sayWorld')
  .arg('user', dl.List(user))
  .arg('color', color)
  .ret(dl.Map(dl.String, user))
  .ok()

it('test user entity', () => {
  const emitter = new EntityEmitter(user, 'ts')
  expect(fmt(emitter.code)).toMatchSnapshot()
})

it('test enum', () => {
  const emitter = new EnumEmitter(color, 'ts')
  expect(fmt(emitter.code)).toMatchSnapshot()
})

it('test user service', () => {
  const serviceEmitter = new ServiceEmitter(userService, 'ts')
  expect(fmt(serviceEmitter.code)).toMatchSnapshot()

  const abstractEmitter = new AbstractService(userService, 'ts')
  expect(fmt(abstractEmitter.code)).toMatchSnapshot()

  const consumerService = new ConsumerEmitter(userService, 'ts')
  expect(fmt(consumerService.code)).toMatchSnapshot()
})

it('test enum service', () => {
  const serviceEmitter = new ServiceEmitter(enumService, 'ts')
  expect(fmt(serviceEmitter.code)).toMatchSnapshot()
})

it('test generic service', () => {
  const serviceEmitter = new AbstractServiceEmitter(gService, 'ts')
  expect(fmt(serviceEmitter.code)).toMatchSnapshot()
})
