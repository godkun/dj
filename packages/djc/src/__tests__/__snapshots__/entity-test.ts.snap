// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test user entity 1`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ 💗 machine coding 💗 ~~~
 */

import java from 'js-to-java'
import s from '@dubbo/sugar'
import Color from '../enum/Color'
import IAddr from '../entity/Addr'

export interface IUser {
  /**
   * 用户id
   */
  id: number
  i: number
  s: number
  us: number
  ch: string
  uch: string

  /**
   * 用户名
   */
  name: string
  bool: boolean
  bl: boolean
  dou: number
  dd: number
  f: number
  ff: number
  ul: number
  l: number
  maps: Map<string>
  ss: Array<string>
  se: Array<Color>
  ll: Array<Color>
  color: Color
  addr: IAddr
  addrs: Array<IAddr>
}

/**
 * 用户模块
 */

export default class User {
  /**
   * 用户id
   */
  id: number
  i: number
  s: number
  us: number
  ch: string
  uch: string

  /**
   * 用户名
   */
  name: string
  bool: boolean
  bl: boolean
  dou: number
  dd: number
  f: number
  ff: number
  ul: number
  l: number
  maps: Map<string>
  ss: Array<string>
  se: Array<Color>
  ll: Array<Color>
  color: Color
  addr: IAddr
  addrs: Array<IAddr>

  constructor(props: IUser) {
    this.id = props.id
    this.i = props.i
    this.s = props.s
    this.us = props.us
    this.ch = props.ch
    this.uch = props.uch
    this.name = props.name
    this.bool = props.bool
    this.bl = props.bl
    this.dou = props.dou
    this.dd = props.dd
    this.f = props.f
    this.ff = props.ff
    this.ul = props.ul
    this.l = props.l
    this.maps = props.maps
    this.ss = props.ss
    this.se = props.se
    this.ll = props.ll
    this.color = props.color
    this.addr = props.addr
    this.addrs = props.addrs
  }

  __fields2java() {
    java('org.apache.dubbo.domain.User', {
      id: java.Integer(this.id),
      i: java.int(this.i),
      s: java.short(this.s),
      us: java.Short(this.us),
      ch: java.char(this.ch),
      uch: java.Char(this.uch),
      name: java.String(this.name),
      bool: java.boolean(this.bool),
      bl: java.Boolean(this.bl),
      dou: java.Double(this.dou),
      dd: java.double(this.dd),
      f: java.float(this.f),
      ff: java.Float(this.ff),
      ul: java.Long(this.ul),
      l: java.long(this.l),
      maps: s.$mhs(this.maps, java.String),
      ss: java.Set(s.$lhs(this.ss, java.String)),
      se: java.Set(s.$lhs(this.se)),
      ll: java.List(s.$lhs(this.ll)),
      color: java.enum('org.apache.dubbo.enum.Color', Color[this.color]),
      addr: this.addr.__fields2java(),
      addrs: java.List(s.$lhs(this.addrs)),
    })
  }
}
"
`;
